In [55]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

#1. Prepare matrix input X and vector label y.

In [56]:
teams = pd.read_csv("teams.csv")
teams

team  year  athletes  events   age  height  weight  prev_medals  medals
0     AFG  1964         8       8  22.0   161.0    64.2          0.0       0
1     AFG  1968         5       5  23.2   170.2    70.0          0.0       0
2     AFG  1972         8       8  29.0   168.3    63.8          0.0       0
3     AFG  1980        11      11  23.6   168.4    63.2          0.0       0
4     AFG  2004         5       5  18.6   170.8    64.8          0.0       0
...   ...   ...       ...     ...   ...     ...     ...          ...     ...
2009  ZIM  2000        26      19  25.0   179.0    71.1          0.0       0
2010  ZIM  2004        14      11  25.1   177.8    70.5          0.0       3
2011  ZIM  2008        16      15  26.1   171.9    63.7          3.0       4
2012  ZIM  2012         9       8  27.3   174.4    65.2          4.0       0
2013  ZIM  2016        31      13  27.5   167.8    62.2          0.0       0

[2014 rows x 9 columns]

In [57]:
# Prepare matrix input X

X = teams[["athletes", "prev_medals"]].copy() 
X["intercept"] = 1                              # Add a column vector with value 1 for bias.
X = X[["intercept", "athletes", "prev_medals"]] # X.shape = (2014, 3)

print(f'X =\n{X}')


X =
      intercept  athletes  prev_medals
0             1         8          0.0
1             1         5          0.0
2             1         8          0.0
3             1        11          0.0
4             1         5          0.0
...         ...       ...          ...
2009          1        26          0.0
2010          1        14          0.0
2011          1        16          3.0
2012          1         9          4.0
2013          1        31          0.0

[2014 rows x 3 columns]


In [58]:
# Prepare vector lebel y

y = teams[["medals"]].copy() # y.shape = (2014, 1)

print(f'y =\n{y}')

y =
      medals
0          0
1          0
2          0
3          0
4          0
...      ...
2009       0
2010       3
2011       4
2012       0
2013       0

[2014 rows x 1 columns]


# 2. Calculate the model using Normal Equation (not using any ML library).
Let B is the vector of coefficients.
The optimal solution is $B = (X^T X)^{-1} X^T y$

For more information, please visit: https://machinelearningcoban.com/2016/12/28/linearregression/

In [59]:
B = np.linalg.inv(X.T @ X) @ X.T @ y
B

medals
0 -1.961889
1  0.071112
2  0.734137

It means that the model is: $y = -1.961889 + 0.071112 x_1 + 0.734137 x_2$

# 3. Calculate the model using Scikit_Learn.

In [60]:
lr = LinearRegression()
lr.fit(teams[["athletes", "prev_medals"]], teams[["medals"]])

LinearRegression()

In [61]:
# Calculate the bias
lr.intercept_

array([-1.96188939])

In [62]:
# Calculate the coeffiecients
lr.coef_

array([[0.07111214, 0.73413679]])

It means that the model is: $y = -1.96188939 + 0.07111214 x_1 + 0.73413679 x_2$